lib/functions/modem.js

```diff
@@ -383,8 +493,91 @@ module.exports = function (SerialPort) {
     }
   }
 
+  modem.checkPINRequired = function (callback, priority, timeout) {
+    if (typeof callback !== 'function') {
+      return new Promise((resolve, reject) => {
+        modem.checkPINRequired((result, error) => {
+          if (error) {
+            reject(error)
+          } else {
+            resolve(result)
+          }
+        }, priority, timeout)
+      })
+    }
+    if (priority == null) priority = false
+    const item = modem.executeCommand('AT+CPIN?', (result, error) => {
+      callback(result, error)
+    }, priority, timeout)
+
+    let resultData
+    item.logic = (newpart) => {
+      if (newpart.startsWith('+CPIN: ')) {
+        resultData = {
+          status: 'success',
+          request: 'checkPINRequired',
+          data: {pinNeeded: !newpart.includes('READY')}
+        }
+      }
+      if (newpart.includes('OK') && resultData) {
```

Comment:
@Apollon77 not all modem returns `OK`. Some only return `+CPIN: READY`

Reply:
really? I aloways get (and found in docs) +CPIN: READY and after that an emoty line and then OK.
If it is really the case as you say then the whole "detect when the command is done" to really send the result is unreliable and we need to change it completely ... But what's the alternative to detect that one command got answered finaly?

Reply:
One idea could be to usilize the current "ECHO" behaviour because (at least for me) all send commands are echoed. So we would need to change the logic so that we check for the appearence of the "ECHO" from the next command in the queue and if this is received without the last queue entry was send we e.g. inject a "virtual" "OK" line (just to finish old command successfully). What do do you think about this?

Reply:
@Apollon77 Yep, I was testing it out, but the initialization was taking too long, and there it is, it didn't go through here:
```js
if (newpart.includes('OK') && resultData) {
        return {
          resultData,
          returnResult: true
        }
      }
```

Yep, the docs says after +CPIN: READY, there would be an 'OK' after. Mine wouldn't. I'm using
these, they're all from China I think:
![IMG_6284](https://user-images.githubusercontent.com/6036299/56863360-f1680a00-69e7-11e9-903b-d1bf24a2ab47.JPG)
and
![IMG_6285](https://user-images.githubusercontent.com/6036299/56863363-f927ae80-69e7-11e9-9d78-da6cbaffed55.JPG)

@Apollon77 How would we also know it it only took a long time for the command to finish, or was there an 'OK' at all.
"But what's the alternative to detect that one command got answered finaly?", I have no idea too, maybe a list of commands and possible results from the modem? An input from the package users, and creating a dictionary out of it?

Reply:
@Apollon77 for the meantime, I'll remove `newpart.includes('OK')` and format the document. And will push and publish after some more checking.

Reply:
But this also may not work because especially here I had lots of problems because then the next command was already send out and was interpreting the "OK" as "OK for himself" and then that whole next commands were broken.

Reply:
What do you think instead of renmove that really checking for "command echos" and internally simulating an OK when next command Echo was received? It feels more "the way to got "if really the devices are that different.
let me try it on my side and PR will come

Reply:
Ok, I will not do an "OK" but an "Onw magic String" like "__COMMAND_END__" that will not really some as data

Reply:
oooh even other idea ... I add a new flag to the item if it is "active" or not and only then the check logic gets executed ... so we wait for the echo nd "activate" the item then for logic processing

Reply:
Are you saying it like this?
```js
      if (newpart.startsWith('+CPIN: ')) {
        resultData = {
          status: 'success',
          request: 'checkPINRequired',
          data: { pinNeeded: !newpart.includes('READY') }
        }
        modem.port.write(`${item.command}OK\r`)
      }
      if (newpart.includes(`${item.command}OK`) && resultData) {
        return {
          resultData,
          returnResult: true
        }
      }
```

Reply:
No ... let me try something ... :-)

Reply:
Please check https://github.com/zabsalahid/serialport-gsm/pull/25 ... maybe that way? :-) Feels sexy to me